File: hospital-ai/backend/app/schemas/appointment.py
"""Appointment domain schemas."""

from datetime import date, datetime, time
from decimal import Decimal
from enum import Enum
from typing import Any, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator


class AppointmentStatus(str, Enum):
    SCHEDULED = "Scheduled"
    COMPLETED = "Completed"
    CANCELLED = "Cancelled"
    NO_SHOW = "No Show"
    RESCHEDULED = "Rescheduled"


class VisitType(str, Enum):
    CONSULTATION = "Consultation"
    FOLLOW_UP = "Follow Up"
    EMERGENCY = "Emergency"
    TELEMEDICINE = "Telemedicine"


ACTIVE_STATUSES = frozenset(
    {
        AppointmentStatus.SCHEDULED,
        AppointmentStatus.RESCHEDULED,
    }
)


class AppointmentAIExtensions(BaseModel):
    """Extension points for future AI agents — currently unused."""

    predicted_wait_time: Optional[int] = None
    priority_score: Optional[Decimal] = None
    recommended_slot: Optional[Any] = None
    ai_notes: Optional[str] = None


class PatientBrief(BaseModel):
    id: UUID
    patient_number: str
    first_name: str
    last_name: str


class DoctorBrief(BaseModel):
    id: UUID
    doctor_number: str
    first_name: str
    last_name: str
    specialization: Optional[str] = None


class DepartmentBrief(BaseModel):
    id: UUID
    name: str


class AppointmentBase(BaseModel):
    patient_id: UUID
    doctor_id: UUID
    department_id: Optional[UUID] = None
    appointment_date: date
    start_time: time
    end_time: time
    visit_type: VisitType = VisitType.CONSULTATION
    reason_for_visit: Optional[str] = Field(default=None, max_length=2000)
    notes: Optional[str] = Field(default=None, max_length=5000)

    @field_validator("reason_for_visit", "notes")
    @classmethod
    def strip_optional(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None

    @model_validator(mode="after")
    def validate_times(self) -> "AppointmentBase":
        if self.end_time <= self.start_time:
            raise ValueError("End time must be after start time")
        start_m = self.start_time.hour * 60 + self.start_time.minute
        end_m = self.end_time.hour * 60 + self.end_time.minute
        duration = end_m - start_m
        if duration < 5:
            raise ValueError("Appointment must be at least 5 minutes")
        if duration > 480:
            raise ValueError("Appointment cannot exceed 8 hours")
        return self


class AppointmentCreate(AppointmentBase):
    status: AppointmentStatus = AppointmentStatus.SCHEDULED


class AppointmentUpdate(BaseModel):
    patient_id: Optional[UUID] = None
    doctor_id: Optional[UUID] = None
    department_id: Optional[UUID] = None
    appointment_date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    status: Optional[AppointmentStatus] = None
    visit_type: Optional[VisitType] = None
    reason_for_visit: Optional[str] = Field(default=None, max_length=2000)
    notes: Optional[str] = Field(default=None, max_length=5000)

    @field_validator("reason_for_visit", "notes")
    @classmethod
    def strip_optional(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None


class AppointmentReschedule(BaseModel):
    appointment_date: date
    start_time: time
    end_time: time
    notes: Optional[str] = Field(default=None, max_length=5000)

    @model_validator(mode="after")
    def validate_times(self) -> "AppointmentReschedule":
        if self.end_time <= self.start_time:
            raise ValueError("End time must be after start time")
        start_m = self.start_time.hour * 60 + self.start_time.minute
        end_m = self.end_time.hour * 60 + self.end_time.minute
        duration = end_m - start_m
        if duration < 5:
            raise ValueError("Appointment must be at least 5 minutes")
        if duration > 480:
            raise ValueError("Appointment cannot exceed 8 hours")
        return self


class AppointmentStatusUpdate(BaseModel):
    status: AppointmentStatus
    notes: Optional[str] = Field(default=None, max_length=5000)


class AppointmentResponse(AppointmentAIExtensions, AppointmentBase):
    id: UUID
    appointment_number: str
    status: AppointmentStatus
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    patient: Optional[PatientBrief] = None
    doctor: Optional[DoctorBrief] = None
    department: Optional[DepartmentBrief] = None


class AppointmentListResponse(BaseModel):
    items: List[AppointmentResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class AvailableSlot(BaseModel):
    start_time: time
    end_time: time


class AvailableSlotsResponse(BaseModel):
    doctor_id: UUID
    appointment_date: date
    day_of_week: int
    available_days: List[int]
    slots: List[AvailableSlot]
    message: Optional[str] = None


class MessageResponse(BaseModel):
    message: str

File: hospital-ai/frontend/src/types/appointment.ts
/** Appointment domain types + AI extension placeholders. */

export type AppointmentStatus =
  | 'Scheduled'
  | 'Completed'
  | 'Cancelled'
  | 'No Show'
  | 'Rescheduled';

export type VisitType =
  | 'Consultation'
  | 'Follow Up'
  | 'Emergency'
  | 'Telemedicine';

/** Reserved for future AI agents — currently unused. */
export interface AppointmentAIExtensions {
  predicted_wait_time: number | null;
  priority_score: number | string | null;
  recommended_slot: unknown | null;
  ai_notes: string | null;
}

export interface PatientBrief {
  id: string;
  patient_number: string;
  first_name: string;
  last_name: string;
}

export interface DoctorBrief {
  id: string;
  doctor_number: string;
  first_name: string;
  last_name: string;
  specialization?: string | null;
}

export interface DepartmentBrief {
  id: string;
  name: string;
}

export interface Appointment extends AppointmentAIExtensions {
  id: string;
  appointment_number: string;
  patient_id: string;
  doctor_id: string;
  department_id: string | null;
  appointment_date: string;
  start_time: string;
  end_time: string;
  status: AppointmentStatus;
  visit_type: VisitType;
  reason_for_visit: string | null;
  notes: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  patient: PatientBrief | null;
  doctor: DoctorBrief | null;
  department: DepartmentBrief | null;
}

export interface AppointmentFormValues {
  patient_id: string;
  doctor_id: string;
  department_id: string;
  appointment_date: string;
  start_time: string;
  end_time: string;
  visit_type: VisitType;
  reason_for_visit: string;
  notes: string;
  status?: AppointmentStatus;
}

export interface AppointmentListParams {
  page?: number;
  page_size?: number;
  search?: string;
  doctor_id?: string;
  patient_id?: string;
  department_id?: string;
  status?: AppointmentStatus | '';
  visit_type?: VisitType | '';
  date_from?: string;
  date_to?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface AppointmentListResponse {
  items: Appointment[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface AvailableSlot {
  start_time: string;
  end_time: string;
}

export interface AvailableSlotsResponse {
  doctor_id: string;
  appointment_date: string;
  day_of_week: number;
  available_days: number[];
  slots: AvailableSlot[];
  message: string | null;
}

export const APPOINTMENT_STATUSES: AppointmentStatus[] = [
  'Scheduled',
  'Completed',
  'Cancelled',
  'No Show',
  'Rescheduled',
];

export const VISIT_TYPES: VisitType[] = [
  'Consultation',
  'Follow Up',
  'Emergency',
  'Telemedicine',
];

export const EMPTY_APPOINTMENT_FORM: AppointmentFormValues = {
  patient_id: '',
  doctor_id: '',
  department_id: '',
  appointment_date: '',
  start_time: '',
  end_time: '',
  visit_type: 'Consultation',
  reason_for_visit: '',
  notes: '',
};

export const STATUS_CHIP_COLORS: Record<AppointmentStatus, string> = {
  Scheduled: 'bg-sky-500/15 text-sky-700 dark:text-sky-300',
  Completed: 'bg-emerald-500/15 text-emerald-700 dark:text-emerald-300',
  Cancelled: 'bg-rose-500/15 text-rose-700 dark:text-rose-300',
  'No Show': 'bg-amber-500/15 text-amber-800 dark:text-amber-300',
  Rescheduled: 'bg-violet-500/15 text-violet-700 dark:text-violet-300',
};

export const CALENDAR_EVENT_COLORS: Record<
  AppointmentStatus | 'Emergency',
  string
> = {
  Scheduled: 'bg-sky-600 border-sky-700',
  Completed: 'bg-emerald-600 border-emerald-700',
  Cancelled: 'bg-rose-600/70 border-rose-700 line-through opacity-80',
  'No Show': 'bg-amber-600 border-amber-700',
  Rescheduled: 'bg-violet-600 border-violet-700',
  Emergency: 'bg-orange-600 border-orange-700',
};

export const DAY_LABELS = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
  'Sunday',
];
